Introduce new command builder DSL

piotrooo opened the pull request:
Below is a new, improved DSL that addresses https://github.com/spring-projects/spring-shell/issues/1127. You’ll find the usage of individual methods (all of them are listed) to make it easier to review everything clearly.

```java
CommandRegistration.builder()
	.command("command")
	.description("description")
	.group("group")

	.interactionMode(InteractionMode.NONINTERACTIVE)
	.isInteractive()
	.isNonInteractive()
	.availability(Availability::available)

	.hidden(true)
	.hidden()

	.withOption(option -> option
			.longNames("arg")
			.shortNames('a')

			.type(String.class)
			.type(ResolvableType.forType(String.class))

			.description("some arg")
			.label("label")
			.defaultValue("defaultValue")
			.nameModifier(String::toUpperCase)

			.required()
			.required(true)

			.position(0)

			.arity(1, 2)
			.arity(OptionArity.ZERO_OR_MORE)

			.completion(completionContext -> List.of())
	)
	.defaultOptionNameModifier(String::toLowerCase)

	.withAlias(alias -> alias
			.command("alias")
			.group("Alias Group")
	)
	.withTarget(target -> target
			.function(function -> function)
			.consumer(commandContext -> {})

			.method(new Object(), "method", String.class)
			.method(new Object(), "method")
	)
	.withExitCode(exitCode -> exitCode
			.map(RuntimeException.class, 1)
			.map(throwable -> 2)
	)
	.withErrorHandling(errorHandling -> errorHandling
			.resolver(ex -> CommandHandlingResult.of("msg"))
	)

	.build();
```

> [!IMPORTANT]
> Since command registration **is the heart** of Spring Shell, it's very important to carefully **review the contracts** of the introduced changes. Modifying this API in the future may be problematic, so **we need to be sure** that this structure is acceptable, as it may stay with us for a long time.

However, when I look at the resulting code — and the existing code in the codebase — I'm not entirely satisfied. The code feels quite clumsy. 😭 

I have a few topics I'd like to bring up.

## Is the naming of these objects good?

I see an analogy here to `BeanDefinition` and `BeanDefinitionRegistry`. Right now, we have the name `CommandRegistration`. But what exactly is this mysterious `Registration`? I have the impression that it's not a good name. It’s not very descriptive and might be misleading.
Maybe we should approach it similarly to the Spring Framework itself and have definitions and registries for commands?

In the definition, we could have a command object with all its metadata — such as names, aliases, and options.

## Should we use interfaces?

I'm not sure what the benefit is of having `CommandRegistration` as an interface (and the rest of objects too). I don't really feel that this approach fits well.

If we had multiple implementations (`1..n`) of different `CommandRegistration`s, then an interface might make sense. Here, it seems unnecessary.

Another question, how should the objects that hold command definition information look? 

Should we use the `Spec` objects everywhere, or only during the build phase — where the result would be an object holder?
For example: `OptionSpec` could create an `OptionDefinition` object (which could be a simple record).

## Aggressiveness of changes

I'm also not sure how far-reaching these changes should be. We're changing the baseline — but are the modifications I’ve introduced acceptable? Could we possibly discuss further, larger changes? Do we need to maintain backward compatibility?

## Summary

Overall, this image perfectly captures how I feel when walking through the classes and trying to understand how commands are created.

<img width="400" height="267" alt="image" src="https://github.com/user-attachments/assets/5764ceb1-3216-4daa-ab07-4d6da1eb94e4" />

> [!CAUTION]
> Without good and clear manual command registration, it will be very difficult to build reliable automatic detection and registration later on.

I'd be happy 🙏 to continue driving this topic, but I need answers to the above questions so I don't go down a dead-end path — and to know how far I can go with the refactor.

piotrooo commented:
> In my opinion, Command should be an interface a user can implement to contribute commands to the shell framework. In a Spring world, we should be able to contribute commands by defining them as beans in the application context:

After some refactoring and playing around with the code, I agree.

> Domain types are now defined as records:

Haha, sorry — this is just how my refactors go. Are you reading my mind? 🤔 Records feel more natural for simple data holders.

> With that in place, and in order to avoid making users implement the entire Command interface, we can define a base implementation of Command with a functional interface (we need to find a good name) to define the command's logic (similar to the Tasklet interface in Spring Batch):

Yes, that syntactic sugar will be super helpful for small tools.

fmbenhassine commented:
Can you please rebase this on the latest `main`? I pushed 97121ac594ba3dc8efe4e0cc3e837caadf49f893 to refine the programming model and which now provides the basic building blocs for the DSL. Sub commands, command availability and aliases are not supported yet, but they will come in 4.0 M3. The support for option arity was removed, I will explain the reason in a separate issue.

So in this first iteration of the DSL, please do include only what is currently supported, and we will add other features in next iterations (make it work, then make it better 😉). I target the initial version of the new DSL in the upcoming 4.0 M2.

Many thanks upfront!

piotrooo commented:
@fmbenhassine I think I have it!

One thing I don’t like is this hacking way of creating the command execution logic:

```java
.execute(commandContext -> {
	String ansiString = new AttributedStringBuilder().append("Good morning ")
		.append("Sir", BOLD.foreground(GREEN))
		.append("!")
		.toAnsi();

	Terminal terminal = commandContext.terminal();
	terminal.writer().println(ansiString);
	terminal.flush();
})
```

I'd prefer to have something like:

```java
.execute(commandContext -> "some string")
```

Without all this complicated flushing and using the terminal.

But we can address this in the next steps, since it’s not easy to define execution precedence. I mean something like:

```java
.execute(commandContext -> {
	String ansiString = new AttributedStringBuilder().append("Good morning ")
		.append("Sir", BOLD.foreground(GREEN))
		.append("!")
		.toAnsi();

	Terminal terminal = commandContext.terminal();
	terminal.writer().println(ansiString);
	terminal.flush();
})
.execute(commandContext -> "some string")
```

Which `execute()` should be executed? But as I said — next steps.

piotrooo commented:
> In the DSL there should be only one call to the execute method (no precedence or ambiguity), and this execute method could/should be overloaded to accept a consumer, a function, etc. With that in place, the user has the choice between the provided options.

This is what I was also thinking about, so to be clear: if these builder definitions are OK for you:

```java
Builder execute(Consumer<CommandContext> commandExecutor);

Builder execute(Function<CommandContext, String> commandExecutor);
```

we can implement them right now.

To prevent using the `execute()` method twice, maybe these methods could return a builder that only has a single `build()` method.

fmbenhassine commented:
> This is what I was also thinking about, so to be clear: if these builder definitions are OK for you:

Yes, that would be great!

> To prevent using the execute() method twice, maybe these methods could return a builder that only has a single build() method.

Exactly, calling `execute` would be the last step in the DSL and would return a builder with a single `build` method that returns the fully defined command. 

Looking forward to reviewing this!

fmbenhassine commented on spring-shell-core/src/main/java/org/springframework/shell/core/command/Command.java:
```diff
@@ -74,4 +76,70 @@ default List<String> getAliases() {
+		 * Set the aliases of the command.
+		 * @return this builder
+		 */
+		Builder aliases(List<String> aliases);
```
Since we have a method accepting aliases as varargs and to keep the API minimalistic, I think there is no need for another one that accepts a list (conversion is always possible on the client side).

fmbenhassine commented on spring-shell-core/src/main/java/org/springframework/shell/core/command/Command.java:
```diff
@@ -74,4 +76,70 @@ default List<String> getAliases() {
+		/**
+		 * Build the {@link AbstractCommand}.
+		 */
+		AbstractCommand build();
```
`AbstractCommand` is a leaky abstraction here (look at the import coming from a different package). The builder should return the interface type (or at least a package private type that we can't break from the outside, but since we don't have that, returning the interface type is the way to go).

fmbenhassine commented on spring-shell-core/src/main/java/org/springframework/shell/core/command/DefaultCommandBuilder.java:
```diff
@@ -0,0 +1,94 @@
+		return abstractCommand;
+	}
+
+	private ConsumerCommandAdapter initCommand() {
```
The code here is used in a single place, so it could be inlined.

piotrooo commented:
@fmbenhassine done. I think I’ve addressed all comments.

One thing that is slightly different is that the `execution()` methods are terminal and return a Command instance.

`EnableCommandRegistrar#registerCommands(Class<?>, BeanDefinitionRegistry)` seems to be unnecessary, but I’ll wait for your opinion.
